// @vizo-o/dev-tools/eslint-config/eslint.config.mjs
import js from "@eslint/js";
import tsPlugin from "@typescript-eslint/eslint-plugin";
import tsParser from "@typescript-eslint/parser";
import prettierPlugin from "eslint-plugin-prettier";
import promisePlugin from "eslint-plugin-promise";
import globals from "globals";

// Clean up globals object to remove any whitespace issues
const browserGlobals = Object.fromEntries(
  Object.entries(globals.browser).map(([key, value]) => [key.trim(), value])
);

const defaultIgnorePatterns = [
  "**/node_modules/**",
  "**/dist/**",
  "**/build/**",
  "**/.next/**",
  "**/coverage/**",
  "**/cdk.out/**",
  "**/*.js",
  "**/*.d.ts",
];

export const createSharedConfig = ({ prettierPlugins = [], isNodeEnv = false,
  ignorePatterns = defaultIgnorePatterns,
  extendIgnorePatterns = [] } = {}) => [
    js.configs.recommended,
    {
      ignores: [
        ...ignorePatterns,
        ...extendIgnorePatterns
      ],
      languageOptions: {
        globals: {
          ...browserGlobals,
          ...globals.es2021,
          ...(isNodeEnv ? globals.node : {}),
        },
      },
    },
    {
      files: ["**/*.{ts,tsx}"],
      languageOptions: {
        parser: tsParser,
        parserOptions: {
          ecmaVersion: "latest",
          sourceType: "module",
          project: "./tsconfig.json",
        },
      },
      plugins: {
        "@typescript-eslint": tsPlugin,
        promise: promisePlugin,
        prettier: prettierPlugin,
      },
      rules: {
        // Prettier integration with inline config
        "prettier/prettier": [
          "warn",
          {
            semi: false,
            singleQuote: true,
            embeddedLanguageFormatting: "auto",
            trailingComma: "all",
            printWidth: 80,
            tabWidth: 4,
            useTabs: false,
            bracketSpacing: true,
            arrowParens: "always",
            endOfLine: "lf",
            bracketSameLine: false,
            jsxSingleQuote: true,
            proseWrap: "preserve",
            quoteProps: "as-needed",
            plugins: prettierPlugins,
          },
        ],

        // TypeScript specific rules
        "no-empty-function": ["error", { "allow": ["constructors"] }],
        "no-template-curly-in-string": "warn",
        "@typescript-eslint/no-floating-promises": "error",
        // Turn off the no-unused-vars rule because it's handled by TypeScript
        "no-unused-vars": "off",
        "@typescript-eslint/no-unused-vars": [
          "error",
          { argsIgnorePattern: "^_", varsIgnorePattern: "^_" },
        ],
        "@typescript-eslint/consistent-type-imports": [
          "error",
          { prefer: "type-imports" },
        ],
        "@typescript-eslint/no-explicit-any": "error",

        // Promise related rules
        "require-await": "error",
        "promise/always-return": "error",
        "promise/no-return-wrap": "error",
        "promise/param-names": "error",
        "promise/catch-or-return": "error",
        "promise/no-new-statics": "error",
        "promise/no-return-in-finally": "error",

        // Modern JavaScript best practices
        "no-var": "error",
        "prefer-const": "error",
        "prefer-arrow-callback": "error",
        "prefer-template": "error",
        eqeqeq: ["error", "always"],
        "space-infix-ops": "error",
        "padding-line-between-statements": [
          "error",
          { blankLine: "always", prev: "*", next: "return" },
        ],
        quotes: [
          "error",
          "single",
          {
            avoidEscape: true,
            allowTemplateLiterals: false,
          },
        ],

        // Turn off formatting rules that Prettier will handle
        semi: "off",
        indent: "off",
      },
    },
  ];

export default createSharedConfig();
